task/clair-scan/0.1/clair-scan.yaml

```diff
@@ -37,12 +67,44 @@ spec:
         #!/usr/bin/env bash
 
         imagewithouttag=$(echo $IMAGE_URL | sed "s/\(.*\):.*/\1/" | tr -d '\n')
-        # strip new-line escape symbol from parameter and save it to variable
-        imageanddigest=$(echo $imagewithouttag@$IMAGE_DIGEST)
+        images_processed_template='{"image": {"pullspec": "'"$IMAGE_URL"'", "digests": [%s]}}'
+        digests_processed=()
+
+        for sha_file in /tekton/home/image-manifest-*.sha; do
+          if [ -e "$sha_file" ]; then
+            arch_sha=$(cat "$sha_file")
+            arch=$(basename "$sha_file" | sed 's/image-manifest-//;s/.sha//')
+            arch_specific_digest="$imagewithouttag@$arch_sha"
+
+            echo "Running clair-action on $arch image manifest."
+            # run the scan for each image manifest in the image index
+            clair-action report --image-ref=$arch_specific_digest --db-path=/tmp/matcher.db --format=quay | tee /tekton/home/clair-result-$arch.json || true
+
+            digests_processed+=("$arch_sha")
+          fi
+        done
 
-        clair-action report --image-ref=$imageanddigest --db-path=/tmp/matcher.db --format=quay | tee /tekton/home/clair-result.json || true
+        # Add the IMAGE_DIGEST to digests_processed array if not present, because even if the image index itself is not scanned,
+        # it gets resolved to amd64 arch which has been scanned.
+        digest_found=0
+        for digest in "${digests_processed[@]}"; do
+          if [[ "$IMAGE_DIGEST" == "$digest" ]]; then
+            digest_found=1
+            break
+          fi
+        done
+
+        if [[ $digest_found -eq 0 ]]; then
+          digests_processed+=("$IMAGE_DIGEST")
```

Comment:
this is like a grey area, I'm not sure if we want to include index image digests in output

@lcarva Does EC checks also index images?

Reply:
Today if an Image Index is provided then EC is just going to check the first architecture. There's planned work to make that better. Eventually EC is going to check each Image Manifest.

I can see an argument here that clair-scan doesn't actually scan the Image Index. It scans each Image Manifest individually and that's what should be reported. At the moment, I don't have a good reason to add the Image Index to the list of images checked. I would suggest not adding it. (We may have to revisit at some point though.)

@arewm, wdyt?

Reply:
I would prefer to leave the Image Index. It isn't scanned because there isn't any content to it other than references, therefore the scan of the reference itself is a no-op. I see it being beneficial to include because this is the image digest that many people will primarily see.

Reply:
My concern with leaving it in is that this creates a special case on the verification side, but it's probably not a big deal. Let's leave it in then.

Reply:
What is the special case on the verification side?

Reply:
One of the values is the digest of the Image Index, not an Image Manifest. Not exactly sure how that will come into play though.

Reply:
I sense most of us are leaning towards leaving it out. I will remove the image index digest from the IMAGES_PROCESSED result. Please let me know if you feel otherwise strongly.

Reply:
As @lcarva told me, it is always easier to add something later if needed than it is to take something out.